### PR TITLE
Split tags into their own `tags` table

### PR DESCRIPTION
**Pull request by fquffio**
This PR fixes an issue that caused all categories to be unlinked from an object when trying to associate a tag to it.

This was caused by the fact that the joint table, `object_categories`, was used for both Categories and Tags, thus the ORM deleted all unneeded records, regardless of the fact that they were pointing to a category or to a tag.

This PR refactors Categories and Tags: tags are now split into their own `tags` table, and linked to Objects via `object_tags`. As a result, `CategoriesTagsBaseTable` has been removed.

### CODE REVIEW

File: plugins/BEdita/Core/src/Model/Entity/ObjectTag.php [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+use BEdita\Core\Utility\JsonApiSerializable;
+use Cake\ORM\Entity;
+
+/**
+ * ObjectTag Entity
+ *
+ * @property int $id
+ * @property int $object_id
+ * @property int $tag_id
+ *
```
Comment (didoda) [inline plugins/BEdita/Core/src/Model/Entity/ObjectTag.php]: minor, remove this line (for cake4 phpcs compatibility)

File: plugins/BEdita/Core/src/Model/Entity/ObjectTag.php [hunk 2]
```diff
@@ -0,0 +1,40 @@
+<?php
+/**
+ * BEdita, API-first content management framework
+ * Copyright 2019 ChannelWeb Srl, Chialab Srl
```
Comment (didoda) [inline plugins/BEdita/Core/src/Model/Entity/ObjectTag.php]:
```suggestion
 * Copyright 2022 ChannelWeb Srl, Chialab Srl
```
minor, current year for copyright

File: plugins/BEdita/Core/src/Model/Table/TagsTable.php [hunk 3]
```diff
@@ -61,46 +66,104 @@ public function initialize(array $config): void
...
+    }
+
+    /**
+     * Find tag IDs by their name.
+     *
+     * @param Query $query Query object.
+     * @param array $options Array containing key `names` as a list of strings.
+     * @return Query
+     */
+    protected function findIds(Query $query, array $options)
```
Comment (didoda) [inline plugins/BEdita/Core/src/Model/Table/TagsTable.php]:
```suggestion
    protected function findIds(Query $query, array $options): Query
```
return type

File: plugins/BEdita/Core/tests/Fixture/ObjectTagsFixture.php [hunk 4]
```diff
@@ -0,0 +1,33 @@
+<?php
+/**
+ * BEdita, API-first content management framework
+ * Copyright 2019 ChannelWeb Srl, Chialab Srl
```
Comment (didoda) [inline plugins/BEdita/Core/tests/Fixture/ObjectTagsFixture.php]:
```suggestion
 * Copyright 2022 ChannelWeb Srl, Chialab Srl
```
minor, current year for copyright

File: plugins/BEdita/Core/tests/Fixture/TagsFixture.php [hunk 5]
```diff
@@ -0,0 +1,38 @@
+<?php
+/**
+ * BEdita, API-first content management framework
+ * Copyright 2019 ChannelWeb Srl, Chialab Srl
```
Comment (didoda) [inline plugins/BEdita/Core/tests/Fixture/TagsFixture.php]:
```suggestion
 * Copyright 2022 ChannelWeb Srl, Chialab Srl
```
minor, current year for copyright

**Review by didoda — APPROVED** [review]

File: plugins/BEdita/Core/src/Model/Entity/Tag.php [hunk 6]
```diff
@@ -51,18 +46,5 @@ class Tag extends Entity implements JsonApiSerializable
      */
     protected $_hidden = [
         '_joinData',
```
Comment (batopa) [inline plugins/BEdita/Core/src/Model/Entity/Tag.php:48]: If I'm not wrong `object_tags` doesn't have join data. I think we can remove `$_hidden` property.

File: plugins/BEdita/Core/src/Model/Table/CategoriesTable.php [hunk 7]
```diff
@@ -121,36 +133,66 @@ public function validationDefault(Validator $validator): Validator
...
+                    true
+                );
+            });
+        });
+    }
+
+    /**
+     * Filter only enabled categories.
+     *
+     * @param Query $query Query object
```
Comment (batopa) [inline plugins/BEdita/Core/src/Model/Table/CategoriesTable.php]:
```suggestion
     * @param \Cake\ORM\Query $query Query object
```

File: plugins/BEdita/Core/src/Model/Table/CategoriesTable.php [hunk 8]
```diff
@@ -121,36 +133,66 @@ public function validationDefault(Validator $validator): Validator
...
+     * @return Query
```
Comment (batopa) [inline plugins/BEdita/Core/src/Model/Table/CategoriesTable.php]:
```suggestion
     * @return \Cake\ORM\Query
```

File: plugins/BEdita/Core/src/Model/Table/CategoriesTable.php [hunk 9]
```diff
@@ -172,6 +214,32 @@ public function findType(Query $query, array $options): Query
         });
     }
 
+    /**
+     * Find categories IDs by their name.
+     *
+     * @param Query $query Query object.
```
Comment (batopa) [inline plugins/BEdita/Core/src/Model/Table/CategoriesTable.php]:
```suggestion
     * @param \Cake\ORM\Query $query Query object.
```

File: plugins/BEdita/Core/src/Model/Table/CategoriesTable.php [hunk 10]
```diff
@@ -172,6 +214,32 @@ public function findType(Query $query, array $options): Query
...
+     * @param array $options Array containing key `names` as a list of strings, and `typeId` as an integer.
+     * @return Query
```
Comment (batopa) [inline plugins/BEdita/Core/src/Model/Table/CategoriesTable.php]:
```suggestion
     * @return \Cake\ORM\Query
```

File: plugins/BEdita/Core/src/Model/Table/TagsTable.php [hunk 11]
```diff
@@ -61,46 +66,104 @@ public function initialize(array $config): void
...
+                    true
+                );
+            });
+        });
+    }
+
+    /**
+     * Filter only enabled tags.
+     *
+     * @param Query $query Query object
```
Comment (batopa) [inline plugins/BEdita/Core/src/Model/Table/TagsTable.php]:
```suggestion
     * @param \Cake\ORM\Query $query Query object
```

File: plugins/BEdita/Core/src/Model/Table/TagsTable.php [hunk 12]
```diff
@@ -61,46 +66,104 @@ public function initialize(array $config): void
...
+     * @return Query
```
Comment (batopa) [inline plugins/BEdita/Core/src/Model/Table/TagsTable.php]:
```suggestion
     * @return \Cake\ORM\Query
```

File: plugins/BEdita/Core/src/Model/Table/TagsTable.php [hunk 13]
```diff
@@ -61,46 +66,104 @@ public function initialize(array $config): void
...
+    {
+        return $query->where([
+            $this->aliasField('enabled') => true,
+        ]);
+    }
+
+    /**
+     * Find tag IDs by their name.
+     *
+     * @param Query $query Query object.
```
Comment (batopa) [inline plugins/BEdita/Core/src/Model/Table/TagsTable.php]:
```suggestion
     * @param \Cake\ORM\Query $query Query object.
```

File: plugins/BEdita/Core/src/Model/Table/TagsTable.php [hunk 14]
```diff
@@ -61,46 +66,104 @@ public function initialize(array $config): void
...
+     * @param array $options Array containing key `names` as a list of strings.
+     * @return Query
```
Comment (batopa) [inline plugins/BEdita/Core/src/Model/Table/TagsTable.php]:
```suggestion
     * @return \Cake\ORM\Query
```

**Review by batopa — CHANGES_REQUESTED** [review]

File: plugins/BEdita/Core/src/Model/Entity/Tag.php (hunk 6 above)
Comment (fquffio) [reply plugins/BEdita/Core/src/Model/Entity/Tag.php:48]: Yeah, correct.

```suggestion
```

**Review by batopa — APPROVED** [review]

**Pull request merged**